### PR TITLE
[Proposal] Return Values of Loads, Samples, and Gathers

### PR DESCRIPTION
**Pull request by bogner**

### CODE REVIEW

File: proposals/NNNN-the-resret-type.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+to make some decisions about how the types will map.
+
+[ResRet]: https://github.com/microsoft/DirectXShaderCompiler/blob/main/docs/DXIL.rst#resource-operation-return-types
+[CheckAccessFullyMapped]: https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/checkaccessfullymapped
+
+## Motivation
+
+The LLVM intrinsics that map to specific DXIL operations aren't constrained to
+fit in a stable binary format, so we have a fair amount of flexibility in how
+wedefine them. We want them to be easy to work with both in the sense that
```
Comment (python3kgae) [inline proposals/NNNN-the-resret-type.md:26]: `we define` or `wedefine`?

File: proposals/NNNN-the-resret-type.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+   things like `Buffer<double2>` here.
+3. An anonymous structure of 4-elements (See llvm/llvm-project#106645), or
+   2-elements in the case of doubles.
+
+The trade offs here are that (1) is the simplest representation in IR but the
+most complex to validate and lower to DXIL operations, and (3) is arguably
+messier but very simple to lower to DXIL operations. Option (2) is a middle
+ground and may be the worst of both worlds.
+
+Discuss: Which is better, (1) or (3)?
```
Comment (llvm-beanz) [inline proposals/NNNN-the-resret-type.md:67]: Do you have any context you can provide about how difficult option 1 is to lower?

**Review by llvm-beanz — APPROVED** [review]

**Review by python3kgae — APPROVED** [review]

File: proposals/NNNN-the-resret-type.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+# Return Values of Loads, Samples, and Gathers
+
+* Proposal: [NNNN](NNNN-the-resret-type.md)
+* Author(s): [Justin Bogner](https://github.com/bogner)
+* Status: **Design In Progress**
+* PRs: llvm/llvm-project#104252, llvm/llvm-project#106645
+
+## Introduction
+
+In dxc a load, sampler, or gather operation returns [a `%dx.types.ResRet`
```
Comment (damyanp) [inline proposals/NNNN-the-resret-type.md:12]:
```suggestion
In dxc a load, sample, or gather operation returns [a `%dx.types.ResRet`
```

**Review by damyanp — APPROVED** [review]

File: proposals/NNNN-the-resret-type.md (hunk 2 above)
Comment (bogner) [reply proposals/NNNN-the-resret-type.md:67]: It isn't a lot more complicated than option 2, which you can see in the PR. The only thing that's particularly different is the "return a scalar" case, and that's pretty simple.

**Pull request merged**